### PR TITLE
477 Make tests fail if sanitizers report errors

### LINKED ISSUE #477: Make CI fail if sanitizers find errors
If address or undefined behavior sanitizers report errors, we currently break the program
- for undefined behavior sanitizer: due to `fno-sanitize-recover=all` 
- for undefined address sanitizer: "If a bug is detected, the program will print an error message to stderr and exit with a non-zero exit code"

See: https://github.com/DLR-SC/memilio/blob/main/cpp/CMakeLists.txt#L55-L56

For documentation of sanitizers see https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html or https://clang.llvm.org/docs/AddressSanitizer.html

However, in these cases, googletest does not recognize a test where the sanitizer recently reported an error

```
2022-11-25T16:23:33.8702714Z 
2022-11-25T16:23:33.8702950Z [----------] 7 tests from TestSaveParameters
2022-11-25T16:23:33.8703402Z [ RUN      ] TestSaveParameters.json_single_sim_write_read_compare
2022-11-25T16:23:33.9018905Z [       OK ] TestSaveParameters.json_single_sim_write_read_compare (31 ms)
2022-11-25T16:23:33.9019443Z [ RUN      ] TestSaveParameters.json_uncertain_matrix_write_read_compare
2022-11-25T16:23:33.9134622Z /home/runner/work/memilio/memilio/cpp/build/_deps/jsoncpp-src/src/lib_json/json_reader.cpp:1559:39: runtime error: negation of -9223372036854775808 cannot be represented in type 'long long int'; cast to an unsigned type to negate this value to itself
```

as failed and gave a green "OK" sign in the github UI.

In order to have correct test reporting, we can take https://github.com/google/googletest/pull/3086

### PR DESCRIPTION
**Pull request by mknaranja**
## Merge Request - GuideLine Checklist 

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
Always to be checked:
* [x] There is at least one issue associated with the pull request.
* [x] The branch follows the naming conventions as defined in the [git workflow](git-workflow).
* [x] New code adheres with the [coding guidelines](coding-guidelines)
* [x] Make sure that the [pre-commit linting/style checks pass](https://github.com/DLR-SC/memilio/wiki).

If functions were changed or functionality was added:
* [x] Tests for new functionality has been added
* [x] A local test was succesful

If new functionality was added:
* [x] There is appropriate **documentation** of your work. (use doxygen style comments)

If new third party software is used:
* [ ] Did you pay attention to its license? Please remember to add it to the wiki after successful merging.

If new mathematical methods or epidemiological terms are used:
* [ ] Are new methods referenced? Did you provide further documentation? Has the glossary been updated? 

The following questions are addressed in the documentation if need be: 
* [ ] Developers (what did you do?, how can it be maintained?)
* [ ] For users (how to use your work?)
* [ ] For admins (how to install and configure your work?)

* For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [ ] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [ ] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [ ] There are appropriate **unit tests** and they pass.
* [ ] The git history is clean and linearized for the merge request. All reviewers should squash commits and write a simple and meaningful commit message.
* [ ] Coverage report for new code is acceptable. 

Closes #477

### CODE REVIEW

**Comment by mknaranja** [comment]
Please note that the CI is supposed to fail as long as #479 is not finalized

File: .github/actions/linux-test/action.yml [hunk 1]
```diff
@@ -29,15 +29,20 @@ runs:
...
+      # always run tests as cmake target coverage
+      # does not return error if a test fail
       run: |
+        cd cpp/build/tests
+        sudo chmod a+x memilio-test
+        ./memilio-test  --gtest_output="xml:report.xml"      
+    - name: Compute code coverage
+      shell: bash
+      # compute code coverage
+      run: |
```
Comment (xsaschako) [inline .github/actions/linux-test/action.yml:41]: I think the logic in the lines afterwards starting with  if-no-files-found: needs to be changed accordingly.

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -52,8 +52,8 @@ if((CMAKE_CXX_COMPILER_ID STREQUAL "GNU") AND(CMAKE_CXX_COMPILER_VERSION VERSION
     endif(MEMILIO_SANITIZE_UNDEFINED)
 
     if(MEMILIO_SANITIZE_ADDRESS OR MEMILIO_SANITIZE_UNDEFINED)
-        string(APPEND CMAKE_CXX_FLAGS_DEBUG " -fno-omit-frame-pointer -fno-sanitize-recover=all")
-        string(APPEND CMAKE_LINKER_FLAGS_DEBUG " -fno-omit-frame-pointer -fno-sanitize-recover=all")
+        string(APPEND CMAKE_CXX_FLAGS_DEBUG " -fno-omit-frame-pointer")
+        string(APPEND CMAKE_LINKER_FLAGS_DEBUG " -fno-omit-frame-pointer")
```
Comment (xsaschako) [inline cpp/CMakeLists.txt:56]: This can be omitted because we implement sanitizers.cpp, am i right?

File: cpp/tests/sanitizers.cpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+}
+void __asan_on_error()
+{
+    FAIL() << "Encountered an address sanitizer error";
+}
+void __tsan_on_report()
+{
+    FAIL() << "Encountered a thread sanitizer error";
+}
+} // extern "C"
```
Comment (xsaschako) [inline cpp/tests/sanitizers.cpp:27]: I don't understand why google tests themselves don't implement this, but okay.

**Review by xsaschako — CHANGES_REQUESTED** [review]
Some minor things and some questions about understand this better, as I don't have that much expertise here. Another discussion for clarification might be helpful.

File: cpp/CMakeLists.txt (hunk 2 above)
Comment (mknaranja) [reply cpp/CMakeLists.txt:56]: no, this is something different.

**Review by xsaschako — APPROVED** [review]

**Pull request merged**